scripts/mapo

```diff
@@ -0,0 +1,231 @@
+        help="The learning rate for the critic (Q-function) optimizer "
+        "(defaul=1e-3).",
+    )
+    optimizers_args_group.add_argument(
+        "--dynamics-lr",
+        type=float,
+        default=1e-3,
+        help="The learning rate for the dynamics optimizer (defaul=1e-3).",
+    )
+    optimizers_args_group.add_argument(
+        "--policy-delay",
+        type=int,
+        default=2,
+        help="The number of traing steps before updating the policy " "(default=2).",
+    )
+    optimizers_args_group.add_argument(
+        "--critic-delay",
+        type=int,
+        default=1,
+        help="The number of traing steps before updating the critic " "(default=2).",
+    )
+
+    exec_args_group = parser.add_argument_group(">> Execution")
+    exec_args_group.add_argument(
+        "--timesteps-total",
+        type=int,
+        default=1000,
+        help="The total number of timesteps (default=1000).",
+    )
+    exec_args_group.add_argument(
+        "--sample-batch-size",
+        type=int,
+        default=10,
+        help="The size of the batch collected by workers (default=10).",
+    )
+    exec_args_group.add_argument(
+        "--train-batch-size",
+        type=int,
+        default=100,
+        help="The size of the batch for training (default=100).",
+    )
+
+    resources_args_group = parser.add_argument_group(">> Resources")
+    resources_args_group.add_argument(
+        "--num-workers",
+        type=int,
+        default=0,
+        help="The number of actors used for parallelism (default=0).",
+    )
+    resources_args_group.add_argument(
+        "--num-gpus",
+        type=float,
+        default=0.0,
+        help="The number of GPUs to allocate to the driver (default=0.0).",
+    )
+    resources_args_group.add_argument(
+        "--num-cpus",
+        type=int,
+        default=1,
+        help="The number of CPUs to allocate to the driver (default=1).",
+    )
+
+    exp_args_group = parser.add_argument_group(">> Experiment")
+    exp_args_group.add_argument(
+        "--num-samples",
+        type=int,
+        default=1,
+        help="The number of times to sample from the hyperparameter space "
+        "(default=1).",
+    )
+    exp_args_group.add_argument("--name", type=str, help="The name of the experiment.")
+
+    return parser.parse_args()
+
+
+def init():
+    """Initialize ray framework."""
+    ray.init()
+
+
+def register():
+    """Register all agents and custom environtments."""
+    register_all_agents()
+    register_all_environments()
+
+
+def run(args):
+    """Run experiments with configuration given by command-line arguments."""
+    trainer = args.run
+    experiment = args.name
+
+    analysis = tune.run(
+        trainer,
+        name=experiment,
+        verbose=args.verbose,
+        stop={"timesteps_total": args.timesteps_total},
+        # resources_per_trial={
+        #     "cpu": args.num_cpus,
+        #     "gpu": args.num_gpus,
+        # },
```

Comment:
Should we keep this comment here? Is it necessary?

Reply:
There's an issue here actually. If I uncomment those I get the following error:

```bash
Traceback (most recent call last):
  File "/home/tbueno/Projects/model-based-reinforcement-learning/bin/mapo", line 7, in <module>
    exec(compile(f.read(), __file__, 'exec'))
  File "/home/tbueno/Projects/model-based-reinforcement-learning/model-aware-policy-optimization/scripts/mapo", line 230, in <module>
    results = run(arguments)
  File "/home/tbueno/Projects/model-based-reinforcement-learning/model-aware-policy-optimization/scripts/mapo", line 200, in run
    "dynamics": args.config_dynamics_net,
  File "/home/tbueno/Projects/model-based-reinforcement-learning/lib/python3.7/site-packages/ray/tune/tune.py", line 244, in run
    runner.step()
  File "/home/tbueno/Projects/model-based-reinforcement-learning/lib/python3.7/site-packages/ray/tune/trial_runner.py", line 323, in step
    next_trial = self._get_next_trial()  # blocking
  File "/home/tbueno/Projects/model-based-reinforcement-learning/lib/python3.7/site-packages/ray/tune/trial_runner.py", line 486, in _get_next_trial
    self._update_trial_queue(blocking=wait_for_trial)
  File "/home/tbueno/Projects/model-based-reinforcement-learning/lib/python3.7/site-packages/ray/tune/trial_runner.py", line 632, in _update_trial_queue
    trials = self._search_alg.next_trials()
  File "/home/tbueno/Projects/model-based-reinforcement-learning/lib/python3.7/site-packages/ray/tune/suggest/basic_variant.py", line 57, in next_trials
    trials = list(self._trial_generator)
  File "/home/tbueno/Projects/model-based-reinforcement-learning/lib/python3.7/site-packages/ray/tune/suggest/basic_variant.py", line 87, in _generate_trials
    experiment_tag=experiment_tag)
  File "/home/tbueno/Projects/model-based-reinforcement-learning/lib/python3.7/site-packages/ray/tune/config_parser.py", line 189, in create_trial_from_spec
    **trial_kwargs)
  File "/home/tbueno/Projects/model-based-reinforcement-learning/lib/python3.7/site-packages/ray/tune/trial.py", line 306, in __init__
    trainable_cls, default_resources))
ValueError: Resources for <class 'ray.rllib.agents.trainer_template.MAPO'> have been automatically set to Resources(cpu=1, gpu=0, extra_cpu=0, extra_gpu=0, custom_resources={}, extra_custom_resources={}) by its `default_resource_request()` method. Please clear the `resources_per_trial` option.
```

Do you have any idea how to fix it?

Reply:
It seems that the `resources_per_trial` option is overridden by the [trainer config](https://github.com/ray-project/ray/blob/e4854b11f9a032261a3a07029b8492e59002da61/python/ray/rllib/agents/trainer.py#L132):
```python
    # === Resources ===
    # Number of actors used for parallelism
    "num_workers": 2,
    # Number of GPUs to allocate to the driver. Note that not all algorithms
    # can take advantage of driver GPUs. This can be fraction (e.g., 0.3 GPUs).
    "num_gpus": 0,
    # Number of CPUs to allocate per worker.
    "num_cpus_per_worker": 1,
    # Number of GPUs to allocate per worker. This can be fractional.
    "num_gpus_per_worker": 0,
    # Any custom resources to allocate per worker.
    "custom_resources_per_worker": {},
    # Number of CPUs to allocate for the driver. Note: this only takes effect
    # when running in Tune.
    "num_cpus_for_driver": 1,
```
So we should pass the command line options as the `num_cpus_for_driver` and `num_gpus`  configurations.

Reply:
I've just tried that and it didn't seem to solve the issue...

Reply:
I tried it with the following command and it worked at the time:
```shell
mapo --run MAPO --env Navigation-v0 --config-actor-net actor.json --config-critic-net critic.json --config-dynamics-net dynamics.json
```

Reply:
Indeed, just needed to merge your feature/on-policy-madpg branch onto my local feature/mapo-running-script.. and issue solved!

Thanks!

Reply:
Sorry, I made a confusion here... actually the problem of the num_gpus and num_cpus persists...

I managed to solve the other issue on the `critic_delay` argument (solved by your commit 39d4d13)... but nevermind we can check this together this week.